src/framework/matrix.hpp

```diff
@@ -533,6 +399,37 @@ template <class T> inline T matrix<T>::operator()(size_t i, size_t j) const {
   return mat_[j * rows_ + i];
 }
 
+// Addressing elements by row or column
+template <class T> inline std::vector<T> matrix<T>::row_index(size_t row) const {
+#ifdef DEBUG
+  if (row >= rows_) {
+    std::cerr << "Error: matrix class operator row_index out of bounds "
+              << row << " >= " << rows_ << std::endl;
+    exit(1);
+  }
+#endif
+  std::vector<T> ret;
+  ret.reserve(cols_);
+  for(size_t i = 0; i < cols_; i++)
+    ret.emplace_back(mat_[i * rows_ + row]);
+  return std::move(ret);
+}
+template <class T> inline std::vector<T> matrix<T>::col_index(size_t col) const {
+#ifdef DEBUG
+  if (col >= cols_) {
+    std::cerr << "Error: matrix class operator col_index out of bounds "
+              << col << " >= " << cols_ << std::endl;
+    exit(1);
+  }
+#endif
+  std::vector<T> ret;
+  ret.reserve(rows_);
+  // we want the elements for all rows i..rows_ and column col
+  for(size_t i = 0; i < rows_; i++)
```

Comment:
Don't change this code (as it's similar to the one in `row_index`), but just wanted to point out that there's an stl algorithm to copy vectors that is more explicit and generates faster code in some situations.
```c++
std::copy_n(mat_.begin() + (col * rows_), rows_, std::back_inserter(ret));
```

Reply:
brilliant thanks - will change here and in getRow